File: 058/main2.c
#include <stdio.h>
#include <math.h>

int diags() {
    static int n = 1;
    static int d = 1;
    static int i = 0;

    n = n + 2 * d;
    if (i == 3) {
        d++;
        i = 0;
    } else {
        i++;
    }

    return n;
}

int isprime(int n) {
    for (int d = 3; d < sqrt(n)+1; d++) {
        if (n % d == 0) {
            return 0;
        }
    }
    return 1;
}

int main () {
    int nprimes = 0;
    int ndiags = 1;
    while (1) {
        for (int i = 0; i < 4; i++) {
            int n = diags();
            ndiags++;
            if (isprime(n)) {
                nprimes++;
            }
        }
        float ratio = nprimes * 1.0 / ndiags;
        /* printf("%d\t%f\n", (ndiags + 1)/2, ratio); */
        if (ratio < 0.1) {
            break;
        }
    }

    printf("%d\n", (ndiags+1)/2);

    return 0;
}

File: 058/main2.py
#!/usr/bin/env python

import  numpy as np

def diags():
    diags.n = diags.n + 2 * diags.d
    if diags.i == 3:
        diags.d += 1
        diags.i = 0
    else:
        diags.i += 1
    return diags.n
diags.n = 1
diags.d = 1
diags.i = 0


def test_diags():
    ds = [diags() for idx in xrange(12)]
    assert ds == [3, 5, 7, 9, 13, 17, 21, 25, 31, 37, 43, 49]

def isprime(n):
    """Check if an odd number (>=3) is a prime"""
    for d in xrange(2, int(np.sqrt(n))+1):
        if n % d == 0:
            return False
    return True


def test_isprime():
    assert isprime(9) == False
    assert isprime(17) == True

if __name__ == '__main__':
    nprimes = 0
    ndiags = 1
    while True:
        for i in range(4):
            n = diags()
            ndiags += 1
            if isprime(n):
                nprimes += 1
        ratio = nprimes * 1.0 / ndiags
        # print (ndiags + 1) / 2, ratio
        if ratio < 0.1:
            break

    print (ndiags + 1) / 2
